Improvements

tawek opened the pull request:
This PR is collection of improvements I've done in order to support Luna HSM (avoiding Thales own provider).

Summary of changes:

1. Optimized the way attributes are read by using heuristic buffer sizes. This improves performance, especially for providers like Luna. The `toString` method now only shows relevant attribute data and the buffer logging doesn't dump on buffer too small errors.

2. Added metrics that are intercepted by the NativeProvider proxy. These metrics capture various data for each function, such as total count of attempts, count of attempts per result, total duration, sum of duration per result, and count of runtime exceptions per exception class.

3. Made `listLen` from the `Template` class accessible. This allows it to be rendered as part of other native structures.

4. Fixed copy and paste for `CKC`.

5. Added more `CKM` mechanisms and other `KDF` and `WRAP` constants. Also added a `CK` class with a couple of common constants.

6. Improved `CKM` by changing `paramLen` in the constructor to `long` to avoid casts and warnings. Now you can pass a `Memory` object directly instead of a pointer and len. The `toString` method now prints `pParameter` instead of `bParameter`.

7. Changed `CKA` bigint operations to use unsigned values as per the PKCS#11 spec.

8. Added a sources jar.

primetomas commented on src/main/java/org/pkcs11/jacknji11/GetAttributeProcess.java:
```diff
@@ -0,0 +1,381 @@
+             Even if some of them are actually invalid we will correct that in next iteration.
+            */
+            bufferTooSmall(response);
+        } else if (rv == CKR.ATTRIBUTE_SENSITIVE || rv == CKR.ATTRIBUTE_TYPE_INVALID) {
```
In my experience it's not that unusual when working with many different HSMs. Different HSMs handle some attributes differently and you can easily get one of these. But it's handled, it was more about the (very good) java comment :-)

tawek replied:
I will remove this "general remark" not to trigger any discussions on its rarity.

primetomas commented on src/main/java/org/pkcs11/jacknji11/GetAttributeProcess.java:
```diff
@@ -0,0 +1,381 @@
+ *
+ * One can change those defaults by setting system properties:
+ * <ul>
+ *     <li>org.pkcs11.jacknji11.GetAttributeProcess.DEFAULT_LARGE_ATTRIBUTE_LENGTH</li>
```
Can we have these settable by code as well. System properties are not ideal when using the library from a larger application, for example running in an application server, inside a container.

primetomas replied:
Oh, and document above (and in setters) as well what setting them to 0 mean.

tawek replied:
If that is your wish then I will make those configurable only via code.  As I see it the configuration is per provider therefore I will move those config options to CryptokiE class.

primetomas commented on src/main/java/org/pkcs11/jacknji11/GetAttributeProcess.java:
```diff
@@ -0,0 +1,381 @@
+                || cka == CKA.COEFFICIENT
+                || cka == CKA.PRIVATE_EXPONENT
+                || cka == CKA.VALUE
+                || cka == CKA.EC_POINT
```
I'd like to see CKA.DATA and CKA.DATA_BITS here. Used for certificates, PQC public keys, and some other objects.

primetomas replied:
A Dilithium5 key (ML-DSA in the future), typically (probably when it's standardized) stored in CKA.DATA is 2592 bytes.
Dilithium2 - 1312 bytes, Dilithium 3 - 1952 bytes, Dilithium 5 - 2592 bytes

tawek replied:
I would gladly add them, except there is no CKA const defined for those.

primetomas replied:
My mistake. CKA.VALUE and CKA_VALUE_BITS. Albeit I must admit I've only seen CKA.VALUE used, so ignore VALUE_BITS. I guess it depends on the use case. CKA_VALUE is used a lot for symmetric keys, key wrapping and such when it's rather small. But also for certificates and larger public keys. 
Would be a trade-off depending on use case wouldn't it?

tawek replied:
I really don't have anything more to contribute here. In context of a stateless GetAttributeValue call VALUE may really be anything (secret or cert). However how frequent you are accessing actual VALUE of a secret? I would say never since it is always SENSITIVE. Therefore VALUE here applies more to certificates than secret keys.

primetomas commented on src/main/java/org/pkcs11/jacknji11/UBigInt.java:
```diff
@@ -0,0 +1,24 @@
+        // utility class
+    }
+
+    public static byte[] ubigint2b(BigInteger value) {
```
Missing javadoc

tawek replied:
OK

primetomas commented on src/main/java/org/pkcs11/jacknji11/Metrics.java:
```diff
@@ -0,0 +1,269 @@
+
+import static java.util.Collections.synchronizedMap;
+
+public class Metrics {
```
Would be great with some class javadoc here

tawek replied:
of course.

primetomas replied:
Assuming there is a performance hit by using Metrics, and you have to enable/disable it?

Very neat feature btw.

tawek replied:
In my world I/O latency triumphs over any microseconds this proxying mechanism might add. Also if you know Spring or Hibernate they basically founded on this. For past 20 years I wasn't complaining about it. Also reflection got a lot faster and is pretty optimized to the point of perfection by now.

primetomas replied:
I've only used debug logging on org.pkcs11.jacknji11.Cryptoki to debug log PKCS#11 calls.
I like the idea of the Metrics, it looks very useful and I think I could make use if it.
Imho it would be better to add Metrics as a separate PR as it is another feature than the attribute fetching. I.e. one feature per PR. What do you think @joelhockey ?

joelhockey replied:
It would be easier to have each of these commits as a separate PR, but it probably doesn't matter too much now

tawek replied:
just 2c: if we split metrics in a separate PR then it must go in before the attribute fetching as I'm checking metrics to verify the optimization in test. I think keeping things separate in logical commits is sufficient and easier to manger and I obviously will squash all the changes from the remarks into target commits just before merging.

joelhockey commented:
Thanks very much @tawek for sending the improvements.  And thanks @primetomas for looking a this - when I looked at the summary of the pull request, my first thought was to ask you to take a look.

primetomas commented:
Thanks. It's a fairly complex PR, so good to have as many eyes as possible. Looks very useful. I've also been looking into optimizing attribute fetches for various HSMs lately (different HSMs behave very differently I noticed).

If one thing, there is lots of indirection, which makes the code clean but harder to understand. It's good with all the javadoc and comments for someone new to understand it in a couple of years. Good java-docing @tawek. Still some more could be added :-).

tawek commented:
Please re-review. I've addressed all remarks above except for "lots of indirection" (sorry). I could address that if you point a finger at a specific "indirection", but I'm satisfied with what it is currently. I also like neat compact code, but the PKCS#11 attribute batch fetching is a little bit of mess to get it right.

primetomas commented on src/main/java/org/pkcs11/jacknji11/GetAttributeLengthStrategy.java:
```diff
@@ -0,0 +1,113 @@
+                CKA.PRIVATE_EXPONENT,
+                CKA.VALUE,
+                CKA.EC_POINT,
+// potentially also these should be added (constants are currently undefined yet)
```
Yes, I was wrong here. Remove.

tawek replied:
OK

primetomas commented:
> Please re-review. I've addressed all remarks above except for "lots of indirection" (sorry). 

You can safely ignore the indirection comment :-)

primetomas commented on src/test/java/org/pkcs11/jacknji11/CryptokiTest.java:
```diff
@@ -21,10 +21,14 @@
 import junit.framework.TestCase;
+import org.pkcs11.jacknji11.GetAttributeLengthStrategy.Default;
+
+import static org.pkcs11.jacknji11.CKA.MODULUS;
```
Why the static import?

tawek replied:
Thats just Idea managing imports automatically. Wanna qualify it instead?

tawek replied:
Done.

primetomas replied:
I think qualifying it makes it more consistent. I.e. CKA.PUBLIC_EXPONENT vs MODULUS_BITS.

joelhockey replied:
yes, please remove the static import for consistency

tawek replied:
Done in whole PR except for for CK and CKA classes.

tawek commented on src/main/java/org/pkcs11/jacknji11/CryptokiE.java:
```diff
@@ -1873,4 +1857,45 @@ private byte[] resize(byte[] buf, int newSize) {
+     *
+     * @param getAttributeLengthStrategy the strategy to use for getting the length of an attribute
+     */
+    public void setQueryLengthStrategy(GetAttributeLengthStrategy getAttributeLengthStrategy) {
```
I would not change it per call. In case of multi-threaded apps it would be rather difficult to manage. I think the objective is met (reducing number of fetches) by default implementation pretty well. I'm just following advice to make things configurable and using " .... different settings per instance ...".

tawek replied:
I think two implementations are currently possible:

1. IndefiniteSize (use length=0 for every attr) - this was previous behavior of the library.
2. MaxSize (use some max length depending on attr) which I implemented in a "Default" impl

It is probably up to you to tell me which one should be the actual default. 
And I would rename the "Default" to "MaxSize" since it conveys what it actually does.

primetomas replied:
Today, previous to your PR, IndefiniteSize would be the "default" right? In that case it may be worth keeping that in order to avoid unexpected behavior change. But making it easy to select MaxSize.
Good idea to rename Default imho.

We would have to do extensive testing on MaxSize, on many different cryptographic modules, in order to make it the default safely without chance of regression. It's probably too much work to test out.

I understand the multi-threading issue. 

Just for the sake of discussion. Perhaps something for the future, actually being able to set size values if you know what size the attributes you're looking for should be? Typically I make calls reading say all attributes for an RSA key, and different ones for an EC key, or PQC keys, based on what KEY_TYPE they report. Typically reading one attribute first, and then the rest in a batch. (Trying to speculatively read all different key attributes gets us into CKR_ATTRIBUTE_TYPE_INVALID on some HSMs, and seems quite slow...)

primetomas replied:
Thanks for the detailed discussion. I do like the idea. It's easier to gather the optimization complexity in the library than to re-implement it everywhere. Also the code is nice imho.
I can run basic tests on a range of different HSMs once the review is done. 

This thread can be resolved if @joelhockey agrees, you have the final say on inclusion.

joelhockey commented on src/main/java/org/pkcs11/jacknji11/CKC.java:
```diff
@@ -38,8 +38,8 @@ public class CKC {
     private static final Map<Long, String> L2S = C.createL2SMap(CKS.class);
```
Is this meant to be CKC.class?

tawek replied:
!fixup added

joelhockey commented on src/main/java/org/pkcs11/jacknji11/CKM.java:
```diff
@@ -21,11 +21,13 @@
 
-import com.sun.jna.Memory;
-import com.sun.jna.Pointer;
+import static org.pkcs11.jacknji11.Hex.b2s;
```
nit: I prefer to avoid churn like changing order of imports and adding extra static imports.

tawek replied:
done

joelhockey replied:
Remove static import of Hex.b2s and use Hex.b2s rather than b2s below

tawek replied:
done

joelhockey commented on src/main/java/org/pkcs11/jacknji11/CKM.java:
```diff
@@ -289,6 +314,10 @@ public class CKM {
+    public static final long CKM_AES_KEY_WRAP            = 0x00002109;
+    public static final long CKM_AES_KEY_WRAP_PAD        = 0x0000210a;
+
+
```
single empty line

tawek replied:
done

joelhockey commented on src/main/java/org/pkcs11/jacknji11/CE.java:
```diff
@@ -1582,4 +1582,9 @@ public static byte[] resize(byte[] buf, int newSize) {
+    public static Metrics getMetrics() {
+        return CRYPTOKIE.getMetrics();
+    }
+
```
remove empty line

tawek replied:
done

joelhockey commented on src/main/java/org/pkcs11/jacknji11/Metrics.java:
```diff
@@ -0,0 +1,269 @@
+        }
+        return sb.toString();
+    }
+
```
nit : remove empty line

tawek replied:
done

joelhockey commented on src/main/java/org/pkcs11/jacknji11/UBigInt.java:
```diff
@@ -0,0 +1,24 @@
+package org.pkcs11.jacknji11;
```
Could you add a copyright notice like the other files?  You are welcome to include yourself as the copyright owner, but license the code using the same MIT.  You are also welcome to assign copyright to me if you want to use the exact same lines from other files (I dont' really care), or you could assign copyright to a vague term like "jacknji11 authors".

tawek replied:
done

joelhockey commented on src/main/java/org/pkcs11/jacknji11/CK.java:
```diff
@@ -0,0 +1,20 @@
+package org.pkcs11.jacknji11;
```
add copyright notice

tawek replied:
done

joelhockey commented on src/main/java/org/pkcs11/jacknji11/Metrics.java:
```diff
@@ -0,0 +1,269 @@
+package org.pkcs11.jacknji11;
```
add copyright notice and license

tawek replied:
done

joelhockey commented on src/main/java/org/pkcs11/jacknji11/GetAttributeProcess.java:
```diff
@@ -0,0 +1,381 @@
+package org.pkcs11.jacknji11;
```
add copyright and license

tawek replied:
done

joelhockey commented on src/test/java/org/pkcs11/jacknji11/CKATest.java:
```diff
@@ -0,0 +1,256 @@
+package org.pkcs11.jacknji11;
```
copyright and license

tawek replied:
done

joelhockey reviewed (commented):
Thanks very much Tomasz and Tomas.  I can't say that I understood every single part of the Metrics and the attribute processing, but I think I got the general idea, and it makes sense.

I've left some minor comments.  If @primetomas is happy, then I will give approve once my comments are resolved

tawek commented on src/main/java/org/pkcs11/jacknji11/Metrics.java:
```diff
@@ -0,0 +1,269 @@
+            }
+        }
+        return method.invoke(nativeProvider, args);
+
```
done

tawek commented:
If I receive green light I'll squash all changes since original PR into those 8 commits.

joelhockey commented:
Feel free to squash back into 8 commits

joelhockey commented:
I intended to submit as separate commits, but I think I pressed the wrong buttons and it all went in as a single commit - it doesn't matter too much.

Thanks again

primetomas commented:
I ran some regression tests with the default strategy (no change from before) on SoftHSM, Crypto4A and DPoD. As expected no noticable changes. So it looks good.
Will test with the MaxLengthStrategy as well.

primetomas commented:
Tested MaxLengthStrategy and saw a significant speedup on DPoD and Thales TCT, no noticeable change (neither faster nor slower) on Crypto4A.

tawek commented:
Any plans to release this?

joelhockey commented:
I don't really do releases.  Is there anything you have in mind? I assume
that anyone using the code is pulling the latest from this repo

On Tue, 20 Feb 2024, 9:14 pm Tomasz Wysocki, ***@***.***>
wrote:

> Any plans to release this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/joelhockey/jacknji11/pull/57#issuecomment-1953990334>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABSZMKTTPRORCPIP44I63LYUSAR3AVCNFSM6AAAAABDE5256CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNJTHE4TAMZTGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

tawek commented:
A version bump at least ?

Also someone made an effort here for previous version ( although under different groupId ) here:

https://mvnrepository.com/artifact/org.signserver.jacknji11/jacknji11

primetomas commented:
Yeah, that's ours (my colleagues). I think it was done for our maven builds to work. We've moved to an internal repo for the pipe-line now so those should probably better be removed.
As joel mentions, we tag our builds on a local copy with internal version numbers.

primetomas commented:
@tawek : I am experiencing issue with the MaxLengthStrategy when running against a Luna DPoD and Luna TCT (at least). They seem to behave badly when it comes to CKR_BUFFER_TOO_SMALL. Instead of setting uLValueLen to -1, uLValueLen is left as it was and then it's not re-read.  This can be easily tested by setting DEFAULT_REGULAR_ATTRIBUTE_LENGTH to 8 instead of 72.

I wonder if you experience the same?

Luna TCT returns a DEVICE_ERROR instead of BUFFER_TOO_SMALL, making it even worse.

primetomas commented:
SoftHSM behaves correctly and sets uLValueLen to -1 when returning BUFFER_TO_SMALL.

primetomas commented:
I also think I found a bug in the buffer_too_small handling. 
In GetAttributeProcess there is today: https://github.com/joelhockey/jacknji11/blob/master/src/main/java/org/pkcs11/jacknji11/GetAttributeProcess.java#L264

Where processAvailable is called after processError, but processError actually only sets it up for another round in the while loop. I think it should read:

            if (rv != CKR.OK) {
                processError(rv, query);
            } else {
                // if we are here it means that there was no actual error,
                // and we can extract available attributes
                processAvailable(query);
            }

Do you agree @tawek ?

primetomas commented:
I created https://github.com/joelhockey/jacknji11/pull/58. Since this is pretty complex with many corner cases, I would like more eyes on this, if you can help @tawek.

primetomas commented:
Ok, opened #59 instead so no-one is surprised if it doesn't work for a specific HSM.
